packages/core/http/core-http-server/src/router/scratch_pad.ts

```diff
@@ -0,0 +1,264 @@
+    {} as any;
+
+  /**
+   * EXAMPLE USAGE:
+   */
+
+  const common = {
+    path: '/api/my-plugin/my-route',
+    options: {},
+  };
+
+  newRouter.post(
+    {
+      ...common,
+      version: '1',
+      validate: { body: schema.object({ name: schema.string() }) },
+    },
+    async (ctx, req, res) => res.ok()
+  );
+
+  newRouter.post(
+    {
+      ...common,
+      version: '2',
+      validate: { body: schema.object({ name: schema.string(), lastName: schema.string() }) },
+    },
+    async (ctx, req, res) => res.ok()
+  );
+
+  newRouter.post(
+    {
+      ...common,
+      version: '3',
+      validate: {
+        body: schema.object({
+          name: schema.string(),
+          lastName: schema.string(),
+          age: schema.number(),
+        }),
+      },
+    },
+    async (ctx, req, res) => res.ok()
+  );
+}
+/**
+ * ===================== End first design =====================
+ */
+
+/**
+ * ===================== Second design =====================
+ *
+ * See "Current API". Per route we registration we could:
+ *
+ * Change: Validation, Handler
+ * Keep constant: Method, Path, Optional config
+ *
+ * If only Validation & Handler change across versions it implies a coupling between these two.
+ *
+ * One approach could be to tightly pair the validation and handler in the registrar's
+ * API.
+ */
+{
+  interface Handler<
+    P,
+    Q,
+    B,
+    Context extends RequestHandlerContextBase = RequestHandlerContextBase,
+    Method extends RouteMethod = 'get'
+  > {
+    validate: RouteValidatorFullConfig<P, Q, B> | false;
+    handler: RequestHandler<P, Q, B, Context, Method>;
+  }
+
+  type NewRouteRegistrar2<
+    Method extends RouteMethod = 'get',
+    Context extends RequestHandlerContextBase = RequestHandlerContextBase
+  > = (
+    route: Omit<RouteConfig<unknown, unknown, unknown, Method>, 'validate'>,
+    handlers: {
+      [version in Version]?: Handler<unknown, unknown, unknown, Context, Method>;
+    }
+  ) => void;
+
+  const newRouter: Omit<IRouter, 'post'> & { post: NewRouteRegistrar2 } = {} as any;
+
+  /**
+   * EXAMPLE USAGE:
+   */
+
+  const v1Route: Handler<unknown, unknown, unknown> = {
+    validate: {
+      body: schema.object({
+        name: schema.string(),
+      }),
+    },
+    handler: async (ctx, req, res) => res.ok(),
+  };
+
+  const v2Route: Handler<unknown, unknown, unknown> = {
+    validate: {
+      body: schema.object({
+        name: schema.string(),
+        lastName: schema.string(),
+      }),
+    },
+    handler: async (ctx, req, res) => res.ok(),
+  };
+
+  newRouter.post(
+    {
+      path: '/api/my-plugin/my-route',
+      options: {},
+    },
+    {
+      '1': v1Route,
+      '2': v2Route,
+    }
+  );
+}
+/**
+ * ===================== End second design =====================
+ */
+
+/**
+ * ===================== Third design =====================
+ * We make the assumption that version is decided at the "router" API level.
```

Comment:
One very nice thing about this design is that we don't actually have change or implement anything for the HTTP router interface until it is absolutely necessary (i.e., when some HTTP API actually changes) since the way routes are registered now is already compatible with a V1 endpoint.